chore: Improve PR template UX (#11318)
stackrox/stackrox: .github/pull_request_template.md

### Description

<!-- A detailed explanation of the changes in your PR. Feel free to remove this section if the title of your PR is sufficiently descriptive. -->

> TBD

Reviewer: Optional suggestion:
```suggestion
change me!
```

## Description

> <!--- A detailed explanation of the changes in your PR. Feel free to remove this section if the title of your PR is sufficiently descriptive. -->

Reviewer: Nitpicking
```suggestion
<!-- A detailed explanation of the changes in your PR. Feel free to remove this section if the title of your PR is sufficiently descriptive. -->
```

Reviewer: Well, ignore me. I was not sophisticated enough to know that triple dash could mean something special.
https://stackoverflow.com/questions/4823468/comments-in-markdown#comment17954715_4829998

Author: Nah, maybe I'm being too smart here. So let's avoid surprises for everyone, especially given [GitHub mentions two dashes](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#hiding-content-with-comments). I'll fix this.

## Checklist

> All items *must be* checked:
> - [ ] CI results are inspected
> - [ ] CHANGELOG is updated **OR** not needed
> - [ ] Documentation PR is created and linked above **OR** not needed

Reviewer: I think, @sthadka and @tommartensen might have considered making it required for all checkboxes to be checked in order to merge the PR. Given that GitHub does not support radio buttons, we can have more items and rely on engineers deleting lines. Deleting is easier than striking through.

I mean this:
```suggestion
All three items *must be* checked:
<!-- Remove conflicting items that won't be checked -->

- [ ] CI results are inspected
- [ ] CHANGELOG is updated
- [ ] I believe CHANGELOG update is not needed
- [ ] Documentation PR is created and linked above
- [ ] Documentation is not needed
```

A valid state would be:

> All three items *must be* checked:
> <!-- Remove items that don't apply -->
> 
> - [x] CI results are inspected
> - [x] I believe CHANGELOG update is not needed
> - [x] Documentation PR is created and linked above

A clearly invalid state includes contradicting items checked (I am a teapot):

> All three items *must be* checked:
> <!-- Remove items that don't apply -->
> 
> - [x] CI results are inspected
> - [x] CHANGELOG is updated
> - [x] I believe CHANGELOG update is not needed
> - [x] Documentation PR is created and linked above
> - [x] Documentation is not needed

Reviewer 2: > I think, @sthadka and @tommartensen might have considered making it required for all checkboxes to be checked in order to merge the PR.

Yes, I have a spike workflow that does exactly that. It uses `Shopify/task-list-checker` action and uses `N/A` and `POST-MERGE` to detect items that can be skipped.

> ## Testing

Reviewer: I highly appreciate that you kept `How I validated my change` section and that we'll have it together with this one. Due to deep spiritual reasons, I suggest a slight titles tweak. Also, markdown linters like a blank line after any heading.

```suggestion
## Testing

### Automated testing

```

What the checklist tells is the state of automated tests, hence `Automated testing` subsection.
Another subsection is `How I validated my change` and it's also part of `Testing`.

Author: My motivation was to strip out everything that's not essential to avoid "template fatigue". So to keep the balance, I'll remove the `## Checklist` header 😛.

Reviewer: Maybe titles can be degraded but kept: `##` -> `###` or even `####`?

Author: good idea, let's try this out

> - [ ] added unit, e2e, regression, and/or compatibility tests
> - [ ] modified existing tests
> - [x] **no new tests added**

Reviewer: Similarly here
```suggestion
There must be at least one item and it must be checked:
<!-- Remove item(s) that don't apply. -->

- [ ] added unit, e2e, regression, and/or compatibility tests
- [ ] modified existing tests
- [x] **no new tests added**
```

Reviewer: Could be also expanded into a more verbose list:

```md
There must be at least one item and it must be checked:
<!-- Remove item(s) that don't apply. -->

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests
- [ ] **no automated tests**
```

Feel free to attach JSON snippets, curl commands, screenshots, etc. Apply a simple benchmark: would the information you
provided convince any reviewer or any external reader that you did enough to validate your change.

It is acceptable to assume trust and keep this section light, e.g. as a bullet-point list.

It is acceptable to skip testing in cases when CI is sufficient, or it's a markdown or code comment change only.  
It is acceptable to skip testing in cases when CI is sufficient, or it's a markdown or code comment change only.
It is also acceptable to skip testing for changes that are too taxing to test before merging. In such case you are
responsible for the change after it gets merged which includes reverting, fixing, etc. Make sure you validate the change
ASAP after it gets merged or explain in PR when the validation will be performed.  
ASAP after it gets merged or explain in PR when the validation will be performed.
Explain here why you skipped testing in case you did so.

Have you created automated tests for your change? Explain here which validation activities you did manually and why so.

> ### Reminder for reviewers
> 
> In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
> performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
> performed testing and you are satisfied with it.

Reviewer: I'd keep this. Could be shortened to:

```suggestion
### Reminder for reviewers

As a reviewer, I confirm that I reviewed testing performed by the author and that, by approving this change, I share responsibility with the author for any problems this code may cause.
```

Author: Do you think having this statement will increase the quality or reviewers' engagement? While I do share the sentiment about sharing the responsibility, the way you suggest to formulate it sounds like a clause in some bank contract which is repelling.

Reviewer: If reviewers understand and share the idea, there's no need to write anything. If you agree that reviewers are all over the place, it would make sense to have something short but to the point in order to help them recognize why they do this.

Author: I think a reminder would be good but given our pool of reviewers is rather static, I'm pretty sure their brains learn quickly to disregard static and annoying parts of the PR, like ads on some website. My goal here is to remove as much boilerplate as possible, so the reviewer's brain is constantly stimulated with varying combinations of checked items.

I'd be willing to try this approach, we can always (re-)add an explicit reminder if we'll see it worked better that way

Reviewer: Ok, let's not bother now.

Author: 🙏

## Description

<!-- A detailed explanation of the changes in your PR. Feel free to remove this section if the title of your PR is sufficiently descriptive. -->

> All three items *must be* checked:
> <!-- Remove conflicting items that won't be checked. -->

Reviewer: This now seems like it's part of Description and so I'm starting to mentally prepare myself PRs that look like this:

![image](https://github.com/stackrox/stackrox/assets/537715/85aae09a-c89c-47fb-b717-d8826410002e)

Not great, in my opinion.
I'd suggest re-introducing a section for these items.

I better thing would be to move "CI results" away from here and call this section for documentation. 
```suggestion
## User-facing documentation

Here must be *two* items and both *must be* checked:
<!-- Remove conflicting items that won't be checked. -->
```

> - [ ] CI results are inspected

Reviewer: This would move away from here to...
```suggestion
```

- [ ] CHANGELOG is updated
- [ ] I believe CHANGELOG update is not needed
- [ ] Documentation PR is created and linked above
- [ ] Documentation is not needed

> ## Testing

Reviewer: ```suggestion
## Testing

- [ ] I inspected CI results
```

### Automated testing

There *must be* at least one item and all items *must be* checked:
<!-- Remove item(s) that don't apply and won't be checked. -->

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

> - [ ] **no automated tests**

Reviewer: Nit: let's have this item in a consistent voice with the ones above. Sorry, did not consider myself when suggested it.

```suggestion
- [ ] **I did not contribute to automated tests**
```

### Description

<!-- A detailed explanation of the changes in your PR. Feel free to remove this section if the title of your PR is sufficiently descriptive. -->

change me!

### User-facing documentation

(*must be* 2 items and both *must be* checked)
<!-- Remove conflicting items that won't be checked. -->

> - [ ] CHANGELOG is updated
> - [ ] CHANGELOG update is not needed
> - [ ] Documentation PR is created and linked above
> - [ ] Documentation is not needed

Reviewer 3: ```suggestion
- [ ] `CHANGELOG.md` is updated
- [ ] `CHANGELOG.md` update is not needed
- [ ] documentation PR is created and linked above
- [ ] documentation is not needed
```

Reviewer 4: Maybe even a link to a [`CHANGELOG.md`](https://github.com/stackrox/stackrox/blob/master/CHANGELOG.md)

### Testing

- [ ] inspected CI results

#### Automated testing

(*must be* at least 1 item and all items *must be* checked)
<!-- Remove item(s) that don't apply and won't be checked. -->

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

> - [ ] contributed **no automated tests**

Reviewer 5: Why there is no `- [ ] existing tests are sufficient`?

Reviewer: Maybe because the author and reviewer(s) did not see how it's different from `- [ ] contributed **no automated tests**`.

Have you created automated tests for your change? Explain here which validation activities you did manually and why so.
-->

> TBD

Reviewer: ```suggestion
change me!
```

> #### Automated testing

Reviewer: This looks extra.
```suggestion
```

Author: 🤦‍♂️ sorry for that
